### PR TITLE
add flink-connector 2.0 writer

### PR DESCRIPTION
**Pull request by Nicole00**

### CODE REVIEW

File: connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    public NebulaCatalog(String catalogName, String defaultDatabase, String username, String pwd,
+                         String address) {
+        super(catalogName, defaultDatabase, username, pwd, address);
+        this.hostAndPorts = NebulaUtils.getHostAndPorts(address);
+        this.metaClient = new MetaClient(hostAndPorts);
+    }
+
+    @Override
+    public List<String> listDatabases() throws CatalogException {
+        Map<String, Integer> spaceMap = new HashMap<>();
```
Comment (chenjiebo) [inline connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java:64]: its better to set the init size of hashmap here

File: connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        try {
+            metaClient.connect();
+            List<IdName> spaces = metaClient.getSpaces();
+            for (IdName space : spaces) {
+                spaceMap.put(new String(space.getName()), space.getId().getSpace_id());
+            }
+        } catch (TException e) {
+            LOG.error(String.format("failed to connect meta service vis %s ", address), e);
```
Comment (chenjiebo) [inline connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java:72]: another more convenient way to write this without using `String.format` which formatting may cause type exceptions:
```
LOG.error("failed to connect meta service vis {}", address, e);
```

File: connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            throw new CatalogException("nebula meta service connect failed.", e);
+        }
+        return new ArrayList<>(spaceMap.keySet());
```
Comment (chenjiebo) [inline connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java:75]: why store space_ids in a map and return only keys, why not just use a set

File: connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    @Override
+    public CatalogBaseTable getTable(ObjectPath tablePath) throws TableNotExistException,
+            CatalogException {
+        if (!tableExists(tablePath)) {
+            throw new TableNotExistException(getName(), tablePath);
+        }
+
+        String graphSpace = tablePath.getDatabaseName();
+        String[] typeAndLabel = tablePath.getObjectName().split(NebulaConstant.SPLIT_POINT);
+        String type = typeAndLabel[0];
```
Comment (chenjiebo) [inline connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java:155]: any chance OutOfBound here?

File: connector/src/main/java/org.apache.flink/connector/nebula/connection/NebulaClientOptions.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            this.timeout = timeout;
+            return this;
+        }
+
+        public NebulaClientOptionsBuilder setConnectRetry(int connectRetry) {
+            this.connectRetry = connectRetry;
+            return this;
+        }
+
+        public NebulaClientOptions build() {
```
Comment (chenjiebo) [inline connector/src/main/java/org.apache.flink/connector/nebula/connection/NebulaClientOptions.java:102]: are all these options required, if so, one can invoke `build()` before he sets all these options, which will cause problems if all these options are required, so I think some checks are necessary

File: connector/src/main/java/org.apache.flink/connector/nebula/sink/NebulaBatchOutputFormat.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+    @Override
+    public void configure(Configuration configuration) {
+    }
+
+    /**
+     * prepare all resources
+     */
+    @Override
+    public void open(int i, int i1) throws IOException {
```
Comment (chenjiebo) [inline connector/src/main/java/org.apache.flink/connector/nebula/sink/NebulaBatchOutputFormat.java:58]: unused params `i` and `i1`

File: connector/src/main/java/org.apache.flink/connector/nebula/sink/NebulaBatchExecutor.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.flink.connector.nebula.statement.EdgeExecutionOptions;
+import org.apache.flink.connector.nebula.statement.ExecutionOptions;
+import org.apache.flink.connector.nebula.statement.VertexExecutionOptions;
+import org.apache.flink.connector.nebula.utils.NebulaConstant;
+import org.apache.flink.connector.nebula.utils.VidTypeEnum;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NebulaBatchExecutor<T> {
+    private static final Logger LOG = LoggerFactory.getLogger(NebulaBatchOutputFormat.class);
```
Comment (chenjiebo) [inline connector/src/main/java/org.apache.flink/connector/nebula/sink/NebulaBatchExecutor.java:23]: wrong class name

File: connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java (hunk 3 above)
Comment (Nicole00) [reply connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java:75]: Thanks for your attention, great catch!

File: connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java (hunk 1 above)
Comment (Nicole00) [reply connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java:64]: Yes，it's better to set map's size if we know the number of elements, but spaces' number is unknown here.

File: connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java (hunk 2 above)
Comment (Nicole00) [reply connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java:72]: 👍

File: connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java (hunk 4 above)
Comment (Nicole00) [reply connector/src/main/java/org.apache.flink/connector/nebula/catalog/NebulaCatalog.java:155]: no chance OutOfBound, the check will stop it.

File: connector/src/main/java/org.apache.flink/connector/nebula/connection/NebulaClientOptions.java (hunk 5 above)
Comment (Nicole00) [reply connector/src/main/java/org.apache.flink/connector/nebula/connection/NebulaClientOptions.java:102]: Great idea, graphSpace and tag/edge must be validate.

**Review by darionyaphet — APPROVED** [review]

**Pull request merged**